relay-profiling/src/sample.rs

```diff
@@ -0,0 +1,429 @@
+    locale: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    manufacturer: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    model: Option<String>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+enum Version {
+    #[serde(rename = "1")]
+    V1,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct SampleProfile {
+    version: Version,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    debug_meta: Option<DebugMeta>,
+
+    device: DeviceMetadata,
+    os: OSMetadata,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    runtime: Option<RuntimeMetadata>,
+
+    #[serde(default, skip_serializing_if = "String::is_empty")]
+    environment: String,
+    #[serde(alias = "profile_id")]
+    event_id: EventId,
+    platform: Platform,
+    profile: Profile,
+    release: String,
+    timestamp: DateTime<Utc>,
+
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    transactions: Vec<TransactionMetadata>,
+}
+
+impl SampleProfile {
+    fn valid(&self) -> bool {
+        match self.platform {
+            Platform::Cocoa => {
+                self.os.build_number.is_some()
+                    && self.device.is_emulator.is_some()
+                    && self.device.locale.is_some()
+                    && self.device.manufacturer.is_some()
+                    && self.device.model.is_some()
+            }
+            Platform::Python => self.runtime.is_some(),
+            Platform::Node => self.runtime.is_some(),
+            _ => true,
+        }
+    }
+
+    /// Removes a sample when it's the only sample on its thread
+    fn remove_single_samples_per_thread(&mut self) {
+        let mut sample_count_by_thread_id: HashMap<u64, u32> = HashMap::new();
+
+        for sample in &self.profile.samples {
+            *sample_count_by_thread_id
+                .entry(sample.thread_id)
+                .or_default() += 1;
+        }
+
+        // Only keep data from threads with more than 1 sample so we can calculate a duration
+        sample_count_by_thread_id.retain(|_, count| *count > 1);
+
+        self.profile
+            .samples
+            .retain(|sample| sample_count_by_thread_id.contains_key(&sample.thread_id));
+    }
+
+    fn strip_pointer_authentication_code(&mut self) {
+        self.profile
+            .strip_pointer_authentication_code(&self.platform, &self.device.architecture);
+    }
+}
+
+pub fn expand_sample_profile(payload: &[u8]) -> Result<Vec<Vec<u8>>, ProfileError> {
+    let profile = parse_profile(payload)?;
+
+    if !profile.valid() {
+        return Err(ProfileError::MissingProfileMetadata);
+    }
+
+    let mut items: Vec<Vec<u8>> = Vec::new();
+
+    for transaction in &profile.transactions {
+        let mut new_profile = profile.clone();
+
+        new_profile.transactions.clear();
+        new_profile.transactions.push(transaction.clone());
```

Comment:
It should be possible to consume `profile.transactions` and prevent the `clone()` here.

Reply:
I'm not sure how to do this. I need to be able to copy the profile but `Profile` doesn't implement `Copy` since I have a bunch of `Vec` and `String` in the `struct`. So I explicitely call `clone`. Then, this is what the compiler suggests:
```rust
error[E0382]: borrow of partially moved value: `profile`
   --> relay-profiling/src/sample.rs:205:31
    |
204 |     for transaction in profile.transactions {
    |                        -------------------- `profile.transactions` partially moved due to this implicit call to `.into_iter()`
205 |         let mut new_profile = profile.clone();
    |                               ^^^^^^^^^^^^^^^ value borrowed here after partial move
    |
note: this function takes ownership of the receiver `self`, which moves `profile.transactions`
    = note: partial move occurs because `profile.transactions` has type `Vec<transaction_metadata::TransactionMetadata>`, which does not implement the `Copy` trait
help: consider iterating over a slice of the `Vec<transaction_metadata::TransactionMetadata>`'s content to avoid moving into the `for` loop
    |
204 |     for transaction in &profile.transactions {

Reply:
This is indeed trickier than I thought, but what should work is

```rust
let mut profile = parse_profile(payload)?;

// [...]

// Replace profile.transactions with an empty vector
// (side benefit: makes calling .clear() unnecessary).
let transactions = std::mem::take(&mut profile.transactions);
for transaction in transactions {
    // [...]
    new_profile.profile.samples.retain_mut(|sample| {
        // [...]
    });

    new_profile.transactions.push(transaction);
```